## Day3-Feature engineering

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv("day2_ML_output.csv")
df.head()

,Age,Cholesterol,Heart rate,Diabetes,Family History,Smoking,Obesity,Alcohol Consumption,Exercise Hours Per Week,Diet,...,Physical Activity Days Per Week,Sleep Hours Per Day,Heart Attack Risk (Binary),Blood sugar,CK-MB,Troponin,Heart Attack Risk (Text),Gender,Systolic blood pressure,Diastolic blood pressure
0,0.595506,0.314286,0.047663,0.0,0.0,1.0,0.0,0.0,0.208326,0,...,0.0,0.333333,0,0.227018,0.048229,0.036512,0,3,0.600000,0.534884
1,0.595506,0.096429,0.047663,1.0,1.0,1.0,1.0,1.0,0.752420,1,...,2.0,0.666667,0,0.227018,0.048229,0.036512,0,3,0.574194,0.569767
2,0.595506,0.189286,0.047663,0.0,0.0,1.0,0.0,1.0,0.200998,2,...,4.0,1.000000,0,0.227018,0.048229,0.036512,0,3,0.187097,0.674419
3,0.078652,0.960714,0.071494,1.0,1.0,1.0,1.0,1.0,0.090557,2,...,1.0,0.500000,0,0.227018,0.048229,0.036512,0,3,0.645161,0.593023
4,0.078652,0.792857,0.071494,1.0,0.0,1.0,1.0,0.0,0.601030,2,...,1.0,0.166667,0,0.227018,0.048229,0.036512,0,3,0.251613,0.383721


In [2]:
df.tail()

,Age,Cholesterol,Heart rate,Diabetes,Family History,Smoking,Obesity,Alcohol Consumption,Exercise Hours Per Week,Diet,...,Physical Activity Days Per Week,Sleep Hours Per Day,Heart Attack Risk (Binary),Blood sugar,CK-MB,Troponin,Heart Attack Risk (Text),Gender,Systolic blood pressure,Diastolic blood pressure
9646,0.404494,0.49978,0.091659,0.652554,0.488749,0.902421,0.50016,0.600192,0.50211,3,...,3.501866,0.504621,0,0.081028,0.002029,0.006894,2,0,1.000000,1.000000
9647,0.235955,0.49978,0.049496,0.652554,0.488749,0.902421,0.50016,0.600192,0.50211,3,...,3.501866,0.504621,0,0.084980,0.003500,0.000194,1,1,0.445161,0.186047
9648,0.348315,0.49978,0.088909,0.652554,0.488749,0.902421,0.50016,0.600192,0.50211,3,...,3.501866,0.504621,0,0.330040,0.009540,0.000194,1,1,0.225806,0.302326
9649,0.370787,0.49978,0.067828,0.652554,0.488749,0.902421,0.50016,0.600192,0.50211,3,...,3.501866,0.504621,0,0.197628,0.119858,0.025439,2,1,0.258065,0.453488
9650,0.797753,0.49978,0.084326,0.652554,0.488749,0.902421,0.50016,0.600192,0.50211,3,...,3.501866,0.504621,0,0.156126,0.006237,0.005923,2,1,0.322581,0.313953


### Feature Engineering
##### New features are created from existing variables to improve model performance and uncover hidden relationships in the dataset.

In [3]:
#Health Risk Score(Aggregation):-
df["Health_Risk_Score"]=(df["Smoking"]+df["Obesity"]+df["Diabetes"]+df["Family History"])
df["Health_Risk_Score"]

0       1.000000
1       4.000000
2       1.000000
3       4.000000
4       3.000000
          ...   
9646    2.543884
9647    2.543884
9648    2.543884
9649    2.543884
9650    2.543884
Name: Health_Risk_Score, Length: 9651, dtype: float64

#### Health Risk Score
##### This feature is the sum of multiple binary risk factors:Smoking,Obesity,Diabetes,Family History.
##### Instead of treating each risk factor separately, this combines them into a single indicator of overall health risk. This can help the model detect cumulative effects.
##### Higher values of Health_Risk_Score are expected to correlate with higher heart attack risk.
##### 0 is Low risk, 4 is Very high risk

In [4]:
#Lifestyle Balance Ratio:-
df["Lifestyle_Ratio"]=df["Exercise Hours Per Week"]/(df["Sedentary Hours Per Day"]+1)
df["Lifestyle_Ratio"]

0       0.134297
1       0.439809
2       0.152216
3       0.064062
4       0.457048
          ...   
9646    0.334758
9647    0.334758
9648    0.334758
9649    0.334758
9650    0.334758
Name: Lifestyle_Ratio, Length: 9651, dtype: float64

##### +1 avoids division by zero.
##### This captures the balance between active and inactive lifestyle, which is more meaningful than using each variable alone.
##### Higher ratio - healthier lifestyle - lower heart attack risk.

In [5]:
#Stress × Heart Rate:-
df["Stress_Heart_Interaction"]=df["Stress Level"]*df["Heart rate"]
df["Stress_Heart_Interaction"]

0       0.428964
1       0.142988
2       0.428964
3       0.071494
4       0.285976
          ...   
9646    0.502047
9647    0.271106
9648    0.486986
9649    0.371515
9650    0.461884
Name: Stress_Heart_Interaction, Length: 9651, dtype: float64

##### Multiplication of stress level and heart rate.
##### High stress combined with high heart rate may indicate cardiovascular strain, making this a powerful interaction feature.
###### Higher values may indicate increased heart attack risk.

In [6]:
df.columns

Index(['Age', 'Cholesterol', 'Heart rate', 'Diabetes', 'Family History',
       'Smoking', 'Obesity', 'Alcohol Consumption', 'Exercise Hours Per Week',
       'Diet', 'Previous Heart Problems', 'Medication Use', 'Stress Level',
       'Sedentary Hours Per Day', 'Income', 'BMI', 'Triglycerides',
       'Physical Activity Days Per Week', 'Sleep Hours Per Day',
       'Heart Attack Risk (Binary)', 'Blood sugar', 'CK-MB', 'Troponin',
       'Heart Attack Risk (Text)', 'Gender', 'Systolic blood pressure',
       'Diastolic blood pressure', 'Health_Risk_Score', 'Lifestyle_Ratio',
       'Stress_Heart_Interaction'],
      dtype='object')

### Evaluate Relationship with Target Variable

In [7]:
target="Heart Attack Risk(Binary)"

In [8]:
print(target)
print(df.columns)

Heart Attack Risk(Binary)
Index(['Age', 'Cholesterol', 'Heart rate', 'Diabetes', 'Family History',
       'Smoking', 'Obesity', 'Alcohol Consumption', 'Exercise Hours Per Week',
       'Diet', 'Previous Heart Problems', 'Medication Use', 'Stress Level',
       'Sedentary Hours Per Day', 'Income', 'BMI', 'Triglycerides',
       'Physical Activity Days Per Week', 'Sleep Hours Per Day',
       'Heart Attack Risk (Binary)', 'Blood sugar', 'CK-MB', 'Troponin',
       'Heart Attack Risk (Text)', 'Gender', 'Systolic blood pressure',
       'Diastolic blood pressure', 'Health_Risk_Score', 'Lifestyle_Ratio',
       'Stress_Heart_Interaction'],
      dtype='object')


In [9]:
features = ["Health_Risk_Score", "Lifestyle_Ratio", "Stress_Heart_Interaction"]
target = 'Heart Attack Risk (Binary)'

for feature in features:
    print("=================================")
    print(f"{feature} vs Target:\n")
    
    result = df.groupby(target)[feature].mean()
    print(result)

Health_Risk_Score vs Target:

Heart Attack Risk (Binary)
0    2.545415
1    2.540979
Name: Health_Risk_Score, dtype: float64
Lifestyle_Ratio vs Target:

Heart Attack Risk (Binary)
0    0.345023
1    0.350614
Name: Lifestyle_Ratio, dtype: float64
Stress_Heart_Interaction vs Target:

Heart Attack Risk (Binary)
0    0.280679
1    0.272386
Name: Stress_Heart_Interaction, dtype: float64


#### Health_Risk_Score
##### 0 - 2.545 , 1 - 2.540, Difference=very very small (~0.005), Both groups (0 = no risk, 1 = risk) have almost same average score.This feature is NOT strongly separating the classes.

#### Lifestyle_Ratio
##### 0 - 0.345 , 1 - 0.350, Slightly higher for risk group (unexpected)Ideally, higher lifestyle ratio = healthier → lower risk, But here it's almost same + slightly reversed.Indicates weak or noisy relationship.

#### Stress_Heart_Interaction
##### 0 - 0.276 , 1 - 0.272, Again very small difference, No clear separation between risk groups, Interaction feature is not contributing much.

In [10]:
df.to_csv("day3_ML_output.csv",index=False)